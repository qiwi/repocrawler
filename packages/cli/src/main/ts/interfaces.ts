import { TGerritkitOpts, TVcs } from '@qiwi/repocrawler-common'
import { EDependencyType, TDepsSource, TReportSortOptions } from '@qiwi/repocrawler-reporters'

export enum EReportType {
  USAGE = 'usage',
  TREE = 'tree',
  VERSIONS = 'versions',
}

export type TReporterCliArgs = {
  package: string
  deps?: EDependencyType
  cwd?: string
  range?: string
  report: EReportType
  sort?: TReportSortOptions
  source: TDepsSource
}

export type TCrawlerBaseOpts = {
  vcs: TVcs
  auth: any
  url: string
}

export type TGithubCrawlerOpts = TCrawlerBaseOpts & {
  vcs: 'github'
  auth: string
}

export type TGerritCrawlerOpts = TCrawlerBaseOpts & {
  vcs: 'gerrit'
  auth: TGerritkitOpts['auth']
}

export type TCrawlerOptionalArgs = {
  limitPeriod?: number
  limitCount?: number
  poolSize?: number
  debug?: boolean
  out?: string
  org?: string[]
  path?: string[]
}

export type TCrawlerOptionalArgsWithConfig = TCrawlerOptionalArgs & {
  config: string
}

export type TCrawlerCliOptsWithoutConfig = TCrawlerBaseOpts & TCrawlerOptionalArgs

export type TCrawlerCliOptsWithConfig = TCrawlerOptionalArgsWithConfig |
  TCrawlerOptionalArgsWithConfig & TCrawlerBaseOpts

export type TCrawlerCliOpts = TCrawlerCliOptsWithConfig | TCrawlerCliOptsWithoutConfig

export type TCrawlerOpts = TGithubCrawlerOpts | TGerritCrawlerOpts

export type TCrawlerCliConfig = TCrawlerOptionalArgs & {
  crawlers: TCrawlerOpts[]
}
