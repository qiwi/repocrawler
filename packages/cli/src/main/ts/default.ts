import { TCrawlerOpts } from '@qiwi/repocrawler-common'

export const defaultCrawlerOpts: TCrawlerOpts = {
  ratelimit: {
    period: 2000,
    count: 10
  },
  debug: true,
}
