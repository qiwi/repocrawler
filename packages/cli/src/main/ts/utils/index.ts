import findCacheDir from 'find-cache-dir'
import { existsSync } from 'fs'
import { Agent } from 'https'
import rimraf from 'rimraf'
import { TCrawlerOpts, TRepoCrawler } from '@qiwi/repocrawler-common'
import { createGithubCrawler } from '@qiwi/repocrawler-github'
import { createGerritCrawler } from '@qiwi/repocrawler-gerrit'
import { TCrawlerBaseOpts } from '../interfaces'

export const getTempDir = (): string =>
  findCacheDir({ name: '@qiwi%2Frepocrawler-cli' }) ||
  '.cache/@qiwi%2Frepocrawler-cli'

export const getNonExistingTempDir = (): string => {
  const path = getTempDir()
  if (existsSync(path)) {
    rimraf.sync(path)
  }
  return path
}

export const createCrawler = (
  { vcs, url, auth }: TCrawlerBaseOpts,
  opts: TCrawlerOpts
): TRepoCrawler => {
  if (vcs === 'github') {
    return createGithubCrawler(
      {
        baseUrl: url,
        auth,
        request: {
          agent: new Agent({
            rejectUnauthorized: false,
          }),
        },
      },
      opts)
  }
  if (vcs === 'gerrit') {
    return createGerritCrawler({ baseUrl: url, auth }, opts)
  }
  throw new Error('Unsupported vcs')
}
