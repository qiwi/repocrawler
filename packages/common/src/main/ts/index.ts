import { AnyMap } from '@qiwi/substrate'

export const foo = 'bar'
export const baz: AnyMap = {}
