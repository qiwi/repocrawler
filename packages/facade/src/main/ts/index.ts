export * from 'blank-ts-monorepo-common'
export * from 'blank-ts-monorepo-plugin'
