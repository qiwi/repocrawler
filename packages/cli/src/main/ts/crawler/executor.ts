import { TRepoCrawler } from '@qiwi/repocrawler-common'
import { ILogger } from '@qiwi/substrate'

import { TCrawlerCliOpts } from '../interfaces'
import { createCrawler } from '../utils'
import { getConfig, resolveCrawlerOpts } from '../utils/config'
import { areOptsWithConfig, validateCrawlerCliArgs } from '../utils/validators'

export const launchCrawler = async (params: TCrawlerCliOpts, logger: ILogger = console): Promise<void> => {
  if (params.org?.length === 0) {
    delete params.org // meow returns empty array for undefined multiple flag
  }

  const validatedParams = validateCrawlerCliArgs(params)
  const configFromFile = areOptsWithConfig(validatedParams) ? getConfig(validatedParams.config) : undefined
  const config = resolveCrawlerOpts(validatedParams, configFromFile)
  const { out, org, debug, limitCount, limitPeriod } = config

  const crawlers: TRepoCrawler[] = config.crawlers.map(opts => createCrawler(
    opts,
    { debug, ratelimit: { period: limitPeriod, count: limitCount }}
  ))
  try {
    await Promise.all(crawlers.map(crawler => crawler.fetchRepoInfo(out, org)))
    logger.log(`Crawling has been finished! Results are saved in ${out}`)
  } catch (e) {
    logger.error(e)
  }
}
