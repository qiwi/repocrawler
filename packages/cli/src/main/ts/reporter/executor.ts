import {
  formatOutput,
  getTreeReport,
  getUsageReport,
  getVersionsReport,
  reportComparatorFactory
} from '@qiwi/repocrawler-reporters'
import { ILogger } from '@qiwi/substrate'

import { EReportType, TReporterCliArgs } from '../interfaces'
import { validateReporterCliArgs } from '../utils/validators'

export const makeReport = async (
  params: TReporterCliArgs,
  logger: ILogger = console
): Promise<void> => {
  if (Array.isArray(params.sort)) {
    delete params.sort
  }
  const {
    package: packageNamePattern,
    deps,
    cwd,
    report: reportType,
    range,
    sort,
    source,
  } = params

  validateReporterCliArgs(params)

  const report =
    reportType === EReportType.TREE
      ? await getTreeReport(packageNamePattern, source, deps, cwd)
      : await (reportType === EReportType.USAGE ? getUsageReport : getVersionsReport)(
        packageNamePattern,
        range as string,
        source,
        deps,
        cwd,
      )

  if (sort) {
    const { field } = sort
    const comparator = reportComparatorFactory(sort)
    if (field === 'version') {
      report.data.sort((a: any, b: any) => comparator(a, b))
    } else {
      report.data.sort((a: any, b: any) => comparator(a[field], b[field]))
    }
  }
  logger.log(formatOutput(report))
}
