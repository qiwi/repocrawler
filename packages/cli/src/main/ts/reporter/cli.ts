#!/usr/bin/env node

import meow from 'meow'

import { EReportType } from '../interfaces'
import { getTempDir } from '../utils'
import { makeReport } from './executor'

const cli = meow(
  `
    Usage:
      reporter --package=^typescript$ --depType=dev --reportType=tree --source=package > report.json
      or
      reporter --package=^@types/ --versionRange=">2.0.0" --depType=all --reportType=usage --source=lock > report.json
    Options
      --cwd, path to crawler's results, default to node_modules/.cache/@qiwi-private%2Fjslab-repocrawler
      --package, package name pattern
      --dep-type, one of [all, dev, peer, optional, default]
      --report, one of [usage, tree]
      --range, semver range, required if reportType === usage
      --chart, number of elements to be showed, only for treeReport 
      --sort.field, one of [name, ratio, usageCount, minVersion, maxVersion, version, project, package, commitDate]
      --sort.order, one of [asc, desc], default to asc
      --source, one of [auto, package, lock]
`,
  {
    importMeta: import.meta,
    flags: {
      cwd: {
        type: 'string',
        default: getTempDir(),
      },
      package: {
        type: 'string',
        isRequired: true,
      },
      reportType: {
        type: 'string',
        isRequired: true,
      },
      depType: {
        type: 'string',
        isRequired: (flags) => flags.source === 'package',
      },
      versionRange: {
        type: 'string',
        isRequired: (flags) => flags.reportType === EReportType.USAGE,
      },
      showChart: {
        type: 'number',
      },
      sort: {
        type: 'string',
        isMultiple: true,
      },
      source: {
        type: 'string',
        isRequired: true,
      },
    },
  },
)

makeReport(cli.flags as any)
  .then(() => process.exit(0))
  .catch(e => {
    console.error(e)
    process.exit(1)
  })
