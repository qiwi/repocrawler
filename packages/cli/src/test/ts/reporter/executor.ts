import { EDependencyType } from '@qiwi/repocrawler-reporters'
import { join } from 'path'

import { EReportType, TReporterCliArgs } from '../../../main/ts/interfaces'
import { makeReport } from '../../../main/ts/reporter/executor'

afterEach(() => jest.clearAllMocks())

const cwd = join(__dirname, '..', '..', 'resources', 'package-jsons')

describe('makeReport', () => {
  it('calls usage report maker and prints report', async () => {
    const args: TReporterCliArgs = {
      report: EReportType.USAGE,
      range: 'foo',
      package: 'bar',
      deps: EDependencyType.ALL,
      source: 'auto',
      cwd,
    }
    jest.spyOn(console, 'log').mockImplementation((data) => {
      expect(data).toMatchSnapshot()
    })
    await makeReport(args)
  })

  it('calls tree report maker and prints report', async () => {
    const args: TReporterCliArgs = {
      report: EReportType.TREE,
      package: 'bar',
      deps: EDependencyType.ALL,
      source: 'auto',
      cwd,
    }
    jest.spyOn(console, 'log').mockImplementation((data) => {
      expect(data).toMatchSnapshot()
    })
    await makeReport(args)
  })

  it('sorts report data', async () => {
    const args = {
      report: EReportType.TREE,
      package: '^@types/jest',
      deps: EDependencyType.ALL,
      cwd,
      source: 'package',
      sort: {
        field: 'usageCount',
        order: 'asc',
      },
    }
    jest.spyOn(console, 'log').mockImplementation((data) => {
      expect(data).toMatchSnapshot()
    })

    await makeReport(args as TReporterCliArgs)
  })
})
