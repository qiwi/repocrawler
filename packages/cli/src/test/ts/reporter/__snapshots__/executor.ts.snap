// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeReport calls tree report maker and prints report 1`] = `
"{
	\\"packageNamePattern\\": \\"bar\\",
	\\"depType\\": \\"all\\",
	\\"source\\": \\"auto\\",
	\\"data\\": [],
	\\"totalProjectsCount\\": 4
}"
`;

exports[`makeReport calls usage report maker and prints report 1`] = `
"{
	\\"packageNamePattern\\": \\"bar\\",
	\\"depType\\": \\"all\\",
	\\"source\\": \\"auto\\",
	\\"versionRange\\": \\"foo\\",
	\\"data\\": []
}"
`;

exports[`makeReport sorts report data 1`] = `
"{
	\\"packageNamePattern\\": \\"^@types/jest\\",
	\\"depType\\": \\"all\\",
	\\"source\\": \\"package\\",
	\\"data\\": [
		{
			\\"name\\": \\"@types/jest-json-schema\\",
			\\"usageCount\\": 3,
			\\"ratio\\": 0.75,
			\\"repos\\": [
				{
					\\"vcs\\": \\"github\\",
					\\"repo\\": \\"package2\\"
				},
				{
					\\"vcs\\": \\"github\\",
					\\"repo\\": \\"package3\\"
				},
				{
					\\"vcs\\": \\"github\\",
					\\"repo\\": \\"package5\\"
				}
			],
			\\"minVersion\\": \\"2.1.2\\",
			\\"maxVersion\\": \\"2.1.2\\"
		},
		{
			\\"name\\": \\"@types/jest\\",
			\\"usageCount\\": 4,
			\\"ratio\\": 1,
			\\"repos\\": [
				{
					\\"vcs\\": \\"gerrit\\",
					\\"repo\\": \\"package1\\"
				},
				{
					\\"vcs\\": \\"github\\",
					\\"repo\\": \\"package2\\"
				},
				{
					\\"vcs\\": \\"github\\",
					\\"repo\\": \\"package3\\"
				},
				{
					\\"vcs\\": \\"github\\",
					\\"repo\\": \\"package5\\"
				}
			],
			\\"minVersion\\": \\"26.0.4\\",
			\\"maxVersion\\": \\"26.0.13\\"
		}
	],
	\\"totalProjectsCount\\": 4
}"
`;
